Feature/selenium ab page object

hktest opened the pull request:
TLT-1719: A/B Project: Selenium Page Object Test for A/B Tool
This PR is to build out a page object structure for the A/B tool, along with initial set of tests to validate the tool's functionality.     

@Joy603 @eparker71  
Priority:  Lower.

Joy603 commented on selenium_tests/ab_tool/ab_base_test_case.py:
```diff
@@ -0,0 +1,36 @@
+
```
Remove blank space

hktest replied:
Done.

Joy603 commented on selenium_tests/ab_tool/ab_tests.py:
```diff
@@ -0,0 +1,60 @@
+        abtext = "A/B Testing Dashboard"
+        print element.text
+        print "Working on verifying that the right page loaded..."
+        driver.save_screenshot('page_after_login.png')
```
You can probably delete screen shot here

hktest replied:
Done.

Joy603 commented on selenium_tests/pin/page_objects/pin_page.py:
```diff
@@ -0,0 +1,67 @@
+
+class PinPageLocators(object):
+    # List of WebElements found on Pin Page (Locators)
+    LOGIN_TYPE_XID_RADIO_BUTTON = (By.ID, "compositeAuthenticationSourceType3")
```
We should not be looking for the third radio button, but for the XID radio button.  Look for the XID text instead of the third radio button, which might change.

hktest replied:
Fixed. Modified to locate the XID radio button using by CSS selector instead on attribute value text.   (By.CSS_SELECTOR, "input[value=\"XID\"]")

rascalking commented on selenium_tests/pin/page_objects/pin_page.py:
```diff
@@ -0,0 +1,67 @@
+        the login to return the appropriate page object for the test. If you do override
+        you will need to call supper to invoke the login
+         """
+        print 'base class logging in user: %s' % username
```
Use logger, not print?

hktest replied:
Discussed - will keep the print statements for now for troubleshooting [primarily for the test reporting].  Will add to list to investigate logger and if it makes sense, and refactor later if needed.

rascalking commented on selenium_tests/ab_tool/page_objects/ab_base.py:
```diff
@@ -0,0 +1,39 @@
+
+
```
nitpick: lose the extra whitespace at the top of the page

hktest replied:
Done.

rascalking commented on selenium_tests/ab_tool/page_objects/ab_base.py:
```diff
@@ -0,0 +1,39 @@
+from selenium.webdriver.common.by import By
+from selenium.webdriver.common.keys import Keys
+from selenium.common.exceptions import NoSuchElementException, NoAlertPresentException
+from abc import abstractmethod
```
style: Ideally, we want the abc import first, then a blank line, then the selenium imports.  See the pep8 [section on imports](https://www.python.org/dev/peps/pep-0008/#imports) for a longer explanation of how to organize them.

hktest replied:
Done.  Good resource/feedback -- thanks!

rascalking commented on selenium_tests/ab_tool/page_objects/ab_base.py:
```diff
@@ -0,0 +1,39 @@
+            pass
+
+        
+    #@abstractmethod
```
just remove this

hktest replied:
Done.

rascalking commented on selenium_tests/ab_tool/page_objects/ab_mainpage.py:
```diff
@@ -0,0 +1,40 @@
+
```
remove the leading empty line

hktest replied:
Done.

rascalking commented on selenium_tests/ab_tool/page_objects/ab_mainpage.py:
```diff
@@ -0,0 +1,40 @@
+    def get_title(self):
+        """This gets the title of the page you're on to validate the right page loaded"""
+        element = self.find_element(*Locator1._h1_selector)
+        print 'title = %s' % element.text
```
either use a logger or remove the print altogether

hktest replied:
same comment above re logger.

rascalking commented on selenium_tests/ab_tool/page_objects/ab_mainpage.py:
```diff
@@ -0,0 +1,40 @@
+        create_experiment.click()
+
+
+    def delete_experiment(self):
```
If you have to call `delete_confirm` every time, maybe rename this to `click_to_begin_delete` or something?

hktest replied:
modified and renamed.

rascalking commented on selenium_tests/ab_tool/page_objects/ab_add_experiment_page.py:
```diff
@@ -0,0 +1,63 @@
+
+class AddExperimentPage(ABBasePage):
+    def __init__(self, driver):
+        super(AddExperimentPage, self).__init__(driver)  
```
You can remove this entire function...if you're not doing anything local to the subclass, you don't need to implement `__init__()`...the superclass' `__init__()` will be called for you automatically.

hktest replied:
good catch, removed.

rascalking commented on selenium_tests/ab_tool/page_objects/ab_add_experiment_page.py:
```diff
@@ -0,0 +1,63 @@
+
+#These are actions that can be taken on the page as part of adding a track
+
+    def create_name (self,driver,my_experiment_name):
```
style: spaces after the commas.

rascalking replied:
This is just filling in the experiment name, right?  Maybe call it `set_experiment_name()` instead?

hktest replied:
Done.

rascalking commented on selenium_tests/ab_tool/page_objects/ab_add_experiment_page.py:
```diff
@@ -0,0 +1,63 @@
+"""This is the page object of the Add Experiment Page"""
+from selenium.webdriver.common.by import By
+from selenium.webdriver.common.keys import Keys
+from selenium_tests.ab_tool.page_objects.ab_base import ABBasePage
```
This belongs after the `from selenium.XYZ` imports, separated by a whitespace.

hktest replied:
This is referring to the comments?  If so, done.  Removed to after all the imports.

rascalking commented on selenium_tests/ab_tool/page_objects/ab_add_experiment_page.py:
```diff
@@ -0,0 +1,63 @@
+    def add_tracks (self,driver,newTrackName):
+        """Adding a track name when creating a new experiment"""
+        new_track_name = driver.find_element(*Locator2._new_track_name)
+        new_track_name.send_keys(newTrackName) 
```
Just curious...the other methods all call `.clear()` before `.send_keys()`...is there a specific reason not to do that here, or just an oversight?

hktest replied:
it's an oversight. for the most part, there fields are blank to begin with, so this there isn't really a need to clear, but if that changed... in any case, I've added it in just to be consistent.

rascalking commented on selenium_tests/ab_tool/page_objects/ab_add_experiment_page.py:
```diff
@@ -0,0 +1,63 @@
+
+    def submit_experiment (self,driver):
+        """Submit to create a new experiment, and confirm create in the modal popup window"""
+        submit = driver.find_element(*Locator2._submit_button).click()
```
If you're not doing anything with the return from `.click()`, don't bother storing it in a variable.  Same goes for the line below this one.

hktest replied:
Fixed.  Good point, claiming newbie status here. :)

rascalking commented on selenium_tests/ab_tool/page_objects/ab_add_experiment_page.py:
```diff
@@ -0,0 +1,63 @@
+        """Validates that uniform random checkbox is checked when creating a new experiment"""
+        try:
+            print "Working on adding a new experiment...:  %s" % self._driver.current_url
+            checked = driver.find_element(*Locator2._uniform_checkbox).is_selected()
```
You may want to separate out the "find checkbox, try focusing on tool frame, then find checkbox" logic, then once you know you've got a valid element, you can just `return uniform_random_checkbox.is_selected()`.

hktest replied:
Another great catch. this function should be been ordered "before" submit_experiment.  Test ordering was incorrect so valid element was not found and attempted to use try/catch to resolve.  Fixed the underlying issue, thus removing the need to check/switch frame.  thanks!

rascalking commented on selenium_tests/ab_tool/page_objects/ab_add_experiment_page.py:
```diff
@@ -0,0 +1,63 @@
+    def is_checked(self, driver):
+        """Validates that uniform random checkbox is checked when creating a new experiment"""
+        try:
+            print "Working on adding a new experiment...:  %s" % self._driver.current_url
```
Either us a logger or remove this.

hktest replied:
same comment above re logger.

rascalking commented on selenium_tests/ab_tool/page_objects/ab_add_experiment_page.py:
```diff
@@ -0,0 +1,63 @@
+        confirm = driver.find_element(*Locator2._confirm_create).click()
+
+
+    def is_checked(self, driver):
```
Should probably be called `is_uniform_random_checked()`.

hktest replied:
Done.

rascalking commented on selenium_tests/ab_tool/regression_suite.py:
```diff
@@ -0,0 +1,37 @@
+import HTMLTestRunner
+
+
+dateTimeStamp = time.strftime('%Y%m%d_%H_%M_%S')
```
style nitpick: python usually prefers underscore_case to camelCase for anything not a class name.

hktest replied:
Fixed.

rascalking commented on selenium_tests/ab_tool/regression_suite.py:
```diff
@@ -0,0 +1,37 @@
+
+dateTimeStamp = time.strftime('%Y%m%d_%H_%M_%S')
+buf = file("TestReport" + "_" + dateTimeStamp + ".html", 'wb')
+runner = HTMLTestRunner.HTMLTestRunner(
```
It's usually a good idea to close files when you're done with them.  Python even gives you a nice way to do that syntatcically:

```python
with file(...) as buf:
    do_something_with(buf)
    do_something_else_with(buf)

# once you're out of the with statement, you know buf is closed
```

hktest replied:
Fixed.  Not an iteration, but added buf.close() to close out file as suggested.  thanks!

rascalking commented on selenium_tests/ab_tool/page_objects/ab_base.py:
```diff
@@ -0,0 +1,39 @@
+
+        
+    #@abstractmethod
+    def _validate_page(self, driver):
```
actually, it doesn't look like this is used anywhere.  maybe remove the whole method?

hktest replied:
Done.

rascalking commented on selenium_tests/ab_tool/ab_base_test_case.py:
```diff
@@ -0,0 +1,36 @@
+from selenium import webdriver
+from selenium_tests.base_page_object import BasePageObject
+from selenium_tests.ab_tool.page_objects.pin_page import LoginPage
+from selenium_tests.base_test_case import BaseSeleniumTestCase
```
same comment about organizing imports applies here.  core libraries first, then third-party modules, then our stuff.  in this case, there's no core libraries, so we'd start with a section that has the `django.conf` and `selenium` imports, followed by a section with the `selenium_tests` imports.

hktest replied:
Done.

rascalking commented on selenium_tests/ab_tool/ab_base_test_case.py:
```diff
@@ -0,0 +1,36 @@
+from selenium_tests.base_test_case import BaseSeleniumTestCase
+from django.conf import settings
+
+from selenium.common.exceptions import NoSuchElementException, TimeoutException, NoAlertPresentException
```
I don't see any of these being used.

hktest replied:
Indeed.  Removed.

rascalking commented on selenium_tests/ab_tool/ab_base_test_case.py:
```diff
@@ -0,0 +1,36 @@
+
+    @classmethod
+    def tearDownClass(cls):
+        super(ABBaseTestCase, cls).tearDownClass()
```
Don't need this, if you're not doing anything different than what the parent class does.

hktest replied:
Fixed/deleted.  Good catch - it's handled by parent class.

rascalking commented on selenium_tests/ab_tool/ab_base_test_case.py:
```diff
@@ -0,0 +1,36 @@
+        #instantiate, then login to URL with username and password from settings, if running locally.
+        base_login = LoginPage(driver) # instantiating 
+        base_login.get(cls.BASE_URL)
+        pin_login = base_login.login(cls.USERNAME, cls.PASSWORD) 
```
If we're not doing anything with `pin_login`, we don't need it.

hktest replied:
Fixed.

rascalking commented on selenium_tests/ab_tool/ab_tests.py:
```diff
@@ -0,0 +1,60 @@
+        page = MainPage(driver) #instantiate
+        element = page.get_title()
+        abtext = "A/B Testing Dashboard"
+        print element.text
```
Use a logger or remove the print altogether.

hktest replied:
see previous comment re: logger.

rascalking commented on selenium_tests/ab_tool/ab_tests.py:
```diff
@@ -0,0 +1,60 @@
+from selenium.webdriver.common.by import By
+from selenium.webdriver.common.keys import Keys
+from selenium.webdriver.support.ui import Select, WebDriverWait
+from selenium.common.exceptions import NoSuchElementException, TimeoutException, NoAlertPresentException
```
Same comment as elsewhere about organizing imports.

hktest replied:
Eek. Fixed!

rascalking commented on selenium_tests/ab_tool/ab_tests.py:
```diff
@@ -0,0 +1,60 @@
+        print element.text
+        print "Working on verifying that the right page loaded..."
+        driver.save_screenshot('page_after_login.png')
+        self.assertEqual(element.text, abtext, element.text)
```
Usually, you want the message to be something like "expected X, got Y", not just "Y".

hktest replied:
Fixed. Modified with cleaner message, "Error: Wrong page. Expected page title is '{}' but page title is returning '{}'".format(abtext, element.text))

rascalking commented on selenium_tests/ab_tool/ab_tests.py:
```diff
@@ -0,0 +1,60 @@
+        fill_form.add_tracks(driver,"My New Track 101")
+        fill_form.add_notes(driver,"My notes")
+        fill_form.submit_experiment(driver)
+        fill_form.is_checked(driver)
```
Does this belong before the call to `submit_experiment()`?

hktest replied:
Fixed.

rascalking commented on selenium_tests/ab_tool/ab_tests.py:
```diff
@@ -0,0 +1,60 @@
+        create_experiment = MainPage(driver) #instantiate
+        create_experiment.click_to_create_experiment()
+        fill_form = AddExperimentPage(driver)
+        fill_form.create_name(driver,"MyExperiment")
```
You're passing driver in everywhere, and it should already be a member of AddExperimentPage...you should be able to refer to it from there, rather than passing it in with every method invocation.

hktest replied:
Fixed.  Figured what I did wrong.  Removed passing in driver at every method invocation, except in the cases of instantiation. Also, as a result, cleaned up AddExperimentPage and removed all unnecessary references to driver (and unused variables).  Very helpful feedback - thanks!

rascalking commented on selenium_tests/ab_tool/ab_tests.py:
```diff
@@ -0,0 +1,60 @@
+            WebDriverWait(driver, 10).until(lambda s: s.find_element(*Locator2._experimentname).is_displayed())
+        except TimeoutException:
+            return False
+        self.assertTrue(self, driver.find_element_by_css_selector("[data-selenium-experiment-name='delete_MyExperiment']"))
```
This is asserting that `self == True`, which I don't think is what you want.

hktest replied:
Good catch.  Fixed, and modified to use use locators as well.  self.assertTrue(driver.find_element(*Locator1._delete_button))

rascalking commented on selenium_tests/ab_tool/ab_tests.py:
```diff
@@ -0,0 +1,60 @@
+        """This deletes the experiment, as part of testing delete functionality and test cleanup"""
+        driver = self.driver
+        delete_experiment = MainPage(driver) #instantiate
+        print "Deleting experiment..."
```
same comment about removing print statements or turning them into logging statements.

hktest replied:
see previous notes re: logger. punted for now.

rascalking commented on selenium_tests/ab_tool/ab_tests.py:
```diff
@@ -0,0 +1,60 @@
+        print "Deleting experiment..."
+        delete_experiment.delete_experiment()
+        delete_experiment.delete_confirm()
+        driver.save_screenshot('delete_experiment.png')
```
You probably want to assert something here.

hktest replied:
Fixed with self.assertTrue(delete_experiment.verify_experiment_is_deleted()).  Also cleaned up delete_experiment and delete_confirm, and removed screenshot save.

rascalking commented:
Mostly style/conventions nitpicks.  Only saw one actual bug (the `self.assertTrue(self, ...)` one), and one place missing something (the test with no asserts).  Looks good!

hktest commented:
@rascalking @Joy603 
- Thank you both for your feedback (very valuable, including stylistic feedback)! 
- Addressed all issues except for logger.  Will add to backlog as item to investigate. Thanks!

Joy603 commented:
This looks great.
